nautobot_golden_config/utilities/helper.py

```diff
@@ -139,3 +150,126 @@ def get_device_to_settings_map(queryset):
+    """
+    device = configs.device
+    jinja_env = jinja2.Environment(autoescape=True)
+
+    for name, func in jinja2_convenience_function().items():
+        # Only importing the encrypt helpers as complements to get_secrets filter
+        if name in ["encrypt_type5", "encrypt_type7"]:
+            jinja_env.filters[name] = func
+
+    # Wrapper for get_secret filter that includes user argument to ensure
+    # that secrets are only rendered by authorized users.
+    # To call this method, the view verifies that it's an authenticated request.
+    jinja_env.filters["get_secret_by_secret_group_slug"] = partial(get_secret_by_secret_group_slug, request.user)
+
+    # Defining the starting intended configuration reference
+    # either a config_to_push, or the default configs.intended_config
+    config_to_render = config_to_push or configs.intended_config
+
+    if not config_to_render:
+        return "No reference Intended configuration to render secrets from"
+
+    try:
+        template = jinja_env.from_string(config_to_render)
+    except jinja_errors.TemplateAssertionError as tae:
+        return f"Jinja encountered an TemplateAssertionError: '{tae}'; check the template for correctness"
+
+    settings = get_device_to_settings_map(Device.objects.filter(pk=device.pk))[device.id]
+    _, device_data = graph_ql_query(request, device, settings.sot_agg_query.query)
+
+    try:
+        return template.render(device_data)
+
+    except jinja_errors.UndefinedError as error:
+        return f"Jinja encountered and UndefinedError: {error}, check the template for missing variable definitions.\n"
+    except jinja_errors.TemplateSyntaxError as error:  # Also catches subclass of TemplateAssertionError
+        return (
+            f"Jinja encountered a SyntaxError at line number {error.lineno},"
+            f"check the template for invalid Jinja syntax.\n"
+        )
+    except jinja_errors.TemplateError:  # Catches all remaining Jinja errors
+        return "Jinja encountered an unexpected TemplateError; check the template for correctness\n"
+
+
+def get_config_to_push(configs: models.GoldenConfig, request: HttpRequest):
+    """Renders final configuration push artifact from intended configuration.
+
+    It chains multiple callables to transform an intended configuration into a configuration that can be pushed.
+    Each callable should match the following signature:
+    `my_callable_function(config_to_push: str, configs: models.GoldenConfig, request: HttpRequest)`
+    """
+    config_to_push = ""
+
+    # Available functions to create the final intended configuration to push
+    config_push_callable = [render_secrets]
+    if PLUGIN_CFG.get("config_push_callable"):
+        config_push_callable = PLUGIN_CFG["config_push_callable"]
+
+    # Actual callable subscribed to post processing the intended configuration
+    config_push_subscribed = [render_secrets]
```

Comment:
My thought process was 

```
    config_push_subscribed = ["render_secrets"]
    if PLUGIN_CFG.get("config_push_subscribed"):
        config_push_subscribed = PLUGIN_CFG["config_push_subscribed"]

    for func in config_push_callable:
        if func.__name__ in config_push_subscribed:
            config_to_push = func(config_to_push, configs, request)
    return config_to_push
```

The reason being is in the future if we want to do something different based on something in the request, we could do so.

Reply:
ok, then what you want is (isn't?):

* In the `nautobot_config.py` -> you define you the available callables, in `config_push_callable` list
* In the `nautobot_config.py` -> you define you the default subscribed, in `config_push_subscribed ` list 
* The `get_config_to_push`, could take an argument in the signature to overwrite the `config_push_subscribed` on the call, by default `None`
* The key ordering, to apply processing ,is the one defined in `config_push_subscribed`

Reply:
@itdependsnetworks check this implementation of this workflow: https://github.com/nautobot/nautobot-plugin-golden-config/pull/339/files#diff-aa599fade7e3a41e0a2d88f9f0634c92bbca44d83a08bcfbe9924a07a24c5154R289